ASFAchievementManager/AchievementHandler.cs

```diff
@@ -146,60 +142,45 @@ public sealed class AchievementHandler : ClientMsgHandler {
 			return result;
 		}
 
-		private void SetStat(List<CMsgClientStoreUserStats2.Stats> statsToSet, List<StatData> Stats, StoredResponse storedResponse, int achievementnum, bool set = true) {
-			if (achievementnum < 0 || achievementnum > Stats.Count) {
-				return; //it should never happen
+		private IEnumerable<CMsgClientStoreUserStats2.Stats> GetStatsToSet(List<CMsgClientStoreUserStats2.Stats> statsToSet, StatData statToSet, bool set = true) {
+			if (statToSet == null) {
+				yield break; //it should never happen
 			}
-			CMsgClientStoreUserStats2.Stats? currentstat = statsToSet.Find(stat => stat.stat_id == Stats[achievementnum].StatNum);
+
+			CMsgClientStoreUserStats2.Stats? currentstat = statsToSet.Find(stat => stat.stat_id == statToSet.StatNum);
 			if (currentstat == null) {
 				currentstat = new CMsgClientStoreUserStats2.Stats() {
-					stat_id = Stats[achievementnum].StatNum,
-					//stat_value = storedResponse.Response.stats.Find(stat => stat.stat_id == Stats[achievementnum].StatNum) != null ? storedResponse.Response.stats.Find(stat => stat.stat_id == Stats[achievementnum].StatNum).stat_value : 0
-					stat_value = storedResponse?.Response?.stats?.Find(stat => stat.stat_id == Stats[achievementnum].StatNum)?.stat_value ?? 0
+					stat_id = statToSet.StatNum,
+					stat_value = statToSet.StatValue
 				};
-				statsToSet.Add(currentstat);
+				yield return currentstat;
 			}
 			if (set) {
-				currentstat.stat_value = currentstat.stat_value | ((uint) 1 << Stats[achievementnum].BitNum);
+				currentstat.stat_value = currentstat.stat_value | ((uint) 1 << statToSet.BitNum);
```

Comment:
you need `(uint) 1 << statToSet.BitNum` in both cases. this is nitpicking, but maybe put it in a variable before the if and use a descriptive name. Makes it a bit more readable